File: metalanguage/src/main/java/relations/MLRelation.java
package relations;

import attributes.MLAttribute;
import constraints.MLConstraint;
import constructs.MLConstruct;
import lombok.*;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Builder
@Getter
public class MLRelation extends MLConstruct {
    @Setter
    private MLRelTypes type;
    @Setter
    private boolean unique;
    @Singular
    private List<MLAttribute> attributes;
    @Singular
    private List<MLConstraint> constraints;
    @Setter
    private MLMultiplicity multiplicity;

    public MLRelation(@NonNull String name, @NonNull UUID id, @NonNull UUID prototypeId) {
        super(name, id, prototypeId);
        attributes = new ArrayList<MLAttribute>();
        constraints = new ArrayList<MLConstraint>();
    }
}

File: metalanguage-lib/src/main/kotlin/ru/psu/relations/MLRelation.kt
package ru.psu.relations

import ru.psu.attributes.MLAttribute
import ru.psu.constructs.MLConstruct
import java.util.*
import kotlin.collections.ArrayList

//Класс отношения в MetaLanguage 1.1
class MLRelation(
        parentId:UUID?, //Идентификатор конструкции-родителя
        id: UUID, //Id отношения
        name: String, //Название отношения
        prototypeId: UUID?, //Id прототипа
        innerStructure: UUID?, //Id графа, описывающего внутреннюю структуру
        var type:MLRelTypes, //Тип отношения
        maxCount:Int, //Максимальное число экземпляров
        var multiplicity:MLMultiplicity, //Множественность отношения
        val ports:MutableSet<UUID> //Id связываемых портов
): MLConstruct(parentId, id, name, prototypeId, innerStructure, maxCount) {
    val attributes:MutableList<MLAttribute> = ArrayList() //Атрибуты отношения
}